src/coprocessor/endpoint.rs

```diff
@@ -51,144 +56,190 @@ pub const DEFAULT_REQUEST_MAX_HANDLE_SECS: u64 = 60;
+            ),
+            high_priority_pool: Host::create_executor_pool(
+                "endpoint-high-pool",
+                cfg.end_point_concurrency,
+                cfg.end_point_stack_size.0 as usize,
+                pd_task_sender,
+            ),
+            basic_local_metrics: BasicLocalMetrics::default(),
             max_running_task_count: cfg.end_point_max_tasks,
             batch_row_limit: cfg.end_point_batch_row_limit,
-            pool: ThreadPoolBuilder::new(
-                thd_name!("endpoint-normal-pool"),
-                CopContextFactory { sender: r.clone() },
-            ).thread_count(cfg.end_point_concurrency)
-                .stack_size(cfg.end_point_stack_size.0 as usize)
-                .build(),
-            low_priority_pool: ThreadPoolBuilder::new(
-                thd_name!("endpoint-low-pool"),
-                CopContextFactory { sender: r.clone() },
-            ).thread_count(cfg.end_point_concurrency)
-                .stack_size(cfg.end_point_stack_size.0 as usize)
-                .build(),
-            high_priority_pool: ThreadPoolBuilder::new(
-                thd_name!("endpoint-high-pool"),
-                CopContextFactory { sender: r.clone() },
-            ).thread_count(cfg.end_point_concurrency)
-                .stack_size(cfg.end_point_stack_size.0 as usize)
-                .build(),
+            stream_batch_row_limit: cfg.end_point_stream_batch_row_limit,
             request_max_handle_duration: cfg.end_point_request_max_handle_duration.0,
-        }
+            running_task_count: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+
+    fn create_executor_pool(
+        name_prefix: &str,
+        pool_size: usize,
+        thread_stack_size: usize,
+        _pd_task_sender: FutureScheduler<PdTask>,
+    ) -> ExecutorPool {
+        let cpu_pool = {
+            CpuPoolBuilder::new()
+                .name_prefix(name_prefix)
+                .pool_size(pool_size)
+                .stack_size(thread_stack_size)
+                .create()
+        };
+        ExecutorPool { pool: cpu_pool }
     }
 
+    #[inline]
     fn running_task_count(&self) -> usize {
-        self.pool.get_task_count() + self.low_priority_pool.get_task_count()
-            + self.high_priority_pool.get_task_count()
+        self.running_task_count.load(Ordering::Acquire)
+    }
+
+    fn notify_failed<E: Into<Error> + Debug>(&mut self, e: E, reqs: Vec<RequestTask>) {
+        debug!("failed to handle batch request: {:?}", e);
+        let resp = err_multi_resp(e.into(), reqs.len(), &mut self.basic_local_metrics);
+        for t in reqs {
+            t.on_resp.respond(resp.clone());
+        }
+    }
+
+    #[inline]
+    fn notify_batch_failed<E: Into<Error> + Debug>(&mut self, e: E, batch_id: u64) {
+        let reqs = self.reqs.remove(&batch_id).unwrap();
+        self.notify_failed(e, reqs);
+    }
+
+    fn handle_request(&mut self, snap: Box<Snapshot>, t: RequestTask) {
```

Comment:
Where is `max_running_task_count` check?

Reply:
It's checked at https://github.com/pingcap/tikv/pull/2478/files#diff-49763d722eeb4620a599e6bddecf8f56R605